sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala

```diff
@@ -0,0 +1,579 @@
+        valueState.update(row.value, row.eventTimeTtl.getTime)
+      } else if (ttlMode == TTLMode.EventTimeTTL()) {
+        valueState.update(row.value)
+      } else {
+        valueState.update(row.value, row.ttl)
+      }
+    } else if (row.action == "get_values_in_ttl_state") {
+      val ttlValues = valueState.getValuesInTTLState()
+      ttlValues.foreach { v =>
+        results = OutputEvent(key, -1, isTTLValue = true, ttlValue = v) :: results
+      }
+    }
+
+    results.iterator
+  }
+}
+
+class ValueStateTTLProcessor
+  extends StatefulProcessor[String, InputEvent, OutputEvent]
+  with Logging {
+
+  @transient private var _valueState: ValueStateImplWithTTL[Int] = _
+  @transient private var _ttlMode: TTLMode = _
+
+  override def init(
+      outputMode: OutputMode,
+      timeoutMode: TimeoutMode,
+      ttlMode: TTLMode): Unit = {
+    _valueState = getHandle
+      .getValueState("valueState", Encoders.scalaInt)
+      .asInstanceOf[ValueStateImplWithTTL[Int]]
+    _ttlMode = ttlMode
+  }
+
+  override def handleInputRows(
+      key: String,
+      inputRows: Iterator[InputEvent],
+      timerValues: TimerValues,
+      expiredTimerInfo: ExpiredTimerInfo): Iterator[OutputEvent] = {
+    var results = List[OutputEvent]()
+
+    for (row <- inputRows) {
+      val resultIter = TTLInputProcessFunction.processRow(_ttlMode, row, _valueState)
+      resultIter.foreach { r =>
+        results = r :: results
+      }
+    }
+
+    results.iterator
+  }
+}
+
+case class MultipleValueStatesTTLProcessor(
+    ttlKey: String,
+    noTtlKey: String)
+  extends StatefulProcessor[String, InputEvent, OutputEvent]
+    with Logging {
+
+  @transient private var _valueStateWithTTL: ValueStateImplWithTTL[Int] = _
+  @transient private var _valueStateWithoutTTL: ValueStateImplWithTTL[Int] = _
+  @transient private var _ttlMode: TTLMode = _
+
+  override def init(
+      outputMode: OutputMode,
+      timeoutMode: TimeoutMode,
+      ttlMode: TTLMode): Unit = {
+    _valueStateWithTTL = getHandle
+      .getValueState("valueState", Encoders.scalaInt)
+      .asInstanceOf[ValueStateImplWithTTL[Int]]
+    _valueStateWithoutTTL = getHandle
+      .getValueState("valueState", Encoders.scalaInt)
+      .asInstanceOf[ValueStateImplWithTTL[Int]]
+    _ttlMode = ttlMode
+  }
+
+  override def handleInputRows(
+      key: String,
+      inputRows: Iterator[InputEvent],
+      timerValues: TimerValues,
+      expiredTimerInfo: ExpiredTimerInfo): Iterator[OutputEvent] = {
+    var results = List[OutputEvent]()
+    val state = if (key == ttlKey) {
+      _valueStateWithTTL
+    } else {
+      _valueStateWithoutTTL
+    }
+
+    for (row <- inputRows) {
+      val resultIterator = TTLInputProcessFunction.processRow(_ttlMode, row, state)
+      resultIterator.foreach { r =>
+        results = r :: results
+      }
+    }
+    results.iterator
+  }
+}
+
+class TransformWithStateTTLSuite
```

Comment:
Could we add a suite level comment for this ?

Reply:
Done.